testutil/ciao-down/instance.go

```diff
@@ -0,0 +1,302 @@
+	Hostname      string `yaml:"hostname"`
+}
+
+func newDefaultInstance(ws *workspace, vmType string) *instance {
+	var in instance
+
+	in.Mounts = []mount{
+		{
+			Tag:           "hostgo",
+			SecurityModel: "passthrough",
+			Path:          ws.GoPath,
+		},
+	}
+
+	in.PortMappings = []portMapping{
+		{
+			Host:  10022,
+			Guest: 22,
+		},
+	}
+
+	if vmType == CIAO {
+		in.PortMappings = append(in.PortMappings, portMapping{
+			Host:  3000,
+			Guest: 3000,
+		})
+	}
+
+	return &in
+}
+
+func newInstanceFromFile(ws *workspace) (*instance, error) {
+	in := &instance{}
+
+	err := in.load(ws)
+	if err == nil {
+		return in, nil
+	}
+
+	// Check for legacy state files.
+
+	vmType := CIAO
+	data, err := ioutil.ReadFile(path.Join(ws.instanceDir, "vmtype.txt"))
+	if err == nil {
+		vmType = string(data)
+		if vmType != CIAO && vmType != CLEARCONTAINERS {
+			err := fmt.Errorf("Unsupported vmType %s. Should be one of "+CIAO+"|"+CLEARCONTAINERS, vmType)
+			return nil, err
+		}
+	}
+
+	uiPath := ""
+	data, err = ioutil.ReadFile(path.Join(ws.instanceDir, "ui_path.txt"))
+	if err == nil {
+		uiPath = string(data)
+	}
+
+	in = newDefaultInstance(ws, vmType)
+
+	if uiPath != "" {
+		in.Mounts = append(in.Mounts, mount{
+			Tag:           "hostui",
+			SecurityModel: "mapped",
+			Path:          filepath.Clean(uiPath),
+		})
+	}
+
+	return in, nil
+}
+
+func (in *instance) unmarshall(data []byte) error {
+	err := yaml.Unmarshal(data, in)
+	if err != nil {
+		return fmt.Errorf("Unable to unmarshall instance state : %v", err)
+	}
+
+	for i := range in.Mounts {
+		if err := checkDirectory(in.Mounts[i].Path); err != nil {
+			return fmt.Errorf("Bad mount %s specified: %v",
+				in.Mounts[i].Path, err)
+		}
+	}
+
+	var memDef, cpuDef int
+	if in.MemGiB == 0 || in.CPUs == 0 {
+		memDef, cpuDef = getMemAndCpus()
+		if in.MemGiB == 0 {
+			in.MemGiB = memDef
+		}
+		if in.CPUs == 0 {
+			in.CPUs = cpuDef
+		}
+	}
+
+	var i int
+	for i = 0; i < len(in.PortMappings); i++ {
+		if in.PortMappings[i].Guest == 22 {
+			break
+		}
+	}
+	if i == len(in.PortMappings) {
+		in.PortMappings = append(in.PortMappings,
+			portMapping{
+				Host:  10022,
+				Guest: 22,
+			})
+	}
+
+	return nil
+}
+
+func (in *instance) unmarshallWithTemplate(ws *workspace, data string) error {
+	tmpl, err := template.New("instance-data").Parse(string(data))
+	if err != nil {
+		return fmt.Errorf("Unable to parse instance data template: %v", err)
+	}
+	var buf bytes.Buffer
+	err = tmpl.Execute(&buf, ws)
+	if err != nil {
+		return fmt.Errorf("Unable to execute instance data template: %v", err)
+	}
+	return in.unmarshall(buf.Bytes())
+}
+
+func (in *instance) load(ws *workspace) error {
+	data, err := ioutil.ReadFile(path.Join(ws.instanceDir, "state.yaml"))
+	if err != nil {
+		return fmt.Errorf("Unable to read instance state : %v", err)
+	}
+
+	return in.unmarshall(data)
+}
+
+func (in *instance) save(ws *workspace) error {
+	data, err := yaml.Marshal(in)
```

Comment:
All, done.  The entire workload specification is now stored in .ciao-down/instance.  The contents of the second document might differ from the original if the user has overriden the default values for cpus and memory, but the 1st and 3rd (userdata) documents will be the same,